active/0000-addon-instrumentation-experimental-hooks.md

```diff
@@ -0,0 +1,373 @@
+- Start Date: Wednesday, 14 December 2016
+- RFC PR: 
+- Ember CLI Issue: 
+
+# Summary
+
+Add build and load performance hooks that are available to addons.  This enables
+users to write addons that do things like summarize and report build performance
+information.
+
+see https://github.com/ember-cli/ember-cli/issues/6349 for additional context.
+
+# Motivation
+
+Build performance is important to users.  We want to enable users to:
+
+1. Easily discover which portions of their build are costly;
+2. Be able to summarize and report build information in an addon;
+3. Be able to write addons that analyze build performance instrumentation so
+   that they can more easily help diagnose build performance issues in projects
+   to which they do not have direct access.  This is of particular interest to
+   @ember-cli/core &c.
+
+In order to provide these hooks to enable iteration and experimentation prior to
+making firm commitments to format, this rfc propose to initially expose them as
+experiments (see the experiments section below).
+
+# Detailed design
+
+## Experiments
+
+Experiments live in `lib/experiments/index.js`.  Unlike feature flags, there is
+no need to strip them from production.  Experiments allow us to provide power
+user features that are not fully stable without their resorting to private API
+usage.
+
+Experiments are available only in canary builds.  This is achieved by only
+including `lib/experiements/index.js` in canary, and making it the entry point
+for all experiments.
+
+## Build & Load Instrumentation Hooks
+
+We have already a build instrumentation hook as an
+experiment in https://github.com/ember-cli/ember-cli/pull/6546
+
+The goal of this RFC is:
+
+  1. To make the concept of experiments supported
+  2. To promote this particular experiment to public API
+
+### Enabling Instrumentation
+
+Instrumentation is enabled if either the environment variable `BROCCOLI_VIZ` is
+set to `1` or if `EMBER_CLI_INSTRUMENTATION` is set to `1`.
+
+If `BROCCOLI_VIZ=1` then in addition to instrumentation hooks being invoked, a
+serialized form of the instrumentation information is written to disk, that is
+appropriate for consumption by [broccoli-viz](https://github.com/ember-cli/broccoli-viz) which is the current behaviour.
+
+### Build Instrumentation
+
+#### Hook
+
+An addon that implements `buildInstrumentation` will have this hook invoked when
+instrumentation is enabled.
+
+```js
+module.exports = {
+  name: 'my-great-addon',
+
+  buildInstrumentation(buildInstrumentation) {
```

Comment:
In the case of that example, it doesn't matter if it fails (for any reason) because we will be warning if the hook doesn't get called.  A better example addressing this specific concern is (used in ember-engines):

```js
var ADDON_TREE_CACHING = (function() {
  try {
    return require('ember-cli/lib/experiments').ADDON_TREE_CACHING;
  } catch (e) {
    if (e.code === 'MODULE_NOT_FOUND') {
      return null;
    }

    throw e;
  }
}());
```

In this case, the only error swallowed is that the module is missing (which is what we would expect for older ember-cli versions).